keps/sig-scheduling/5732-topology-aware-workload-scheduling/README.md

```diff
@@ -0,0 +1,990 @@
+   co-located resource (e.g., interconnected network interfaces or
+   accelerators).
+
+The scheduler is extended to interpret these constraints and find a "Placement"
+(a subset of nodes and DRA resources) that satisfies them.
+
+### User Stories (Optional)
+
+#### Story 1: AI Training in a Single Rack
+
+As a data scientist, I want to run a distributed training job where all pods
+need to be located in the same server rack to minimize latency. I define a
+`TopologyConstraint` on the Workload's PodGroup specifying the rack topology
+label. The scheduler identifies a rack with sufficient capacity and schedules
+all pods there at once.
+
+#### Story 2: Workload using Interconnected DRA Devices
+
+As a cluster administrator, I want to schedule a workload that requires a set of
+specialized accelerators that are physically interconnected. I use a
+`DRAConstraint` targeting a specific `ResourceClaimTemplate`. The scheduler
+finds a set of DRA resources (ResourceSlice) that are co-located and binds the
+workload's pods to them.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- **Scheduling Latency:** Evaluating multiple placements involves running
+  filter/score plugins multiple times.
+
+  - **Mitigation:** Implement pre-filtering optimizations to reject infeasible
+    placements early based on aggregate resource availability.
+
+- **Complexity of Pod Group Scheduling:** Scheduling heterogeneous Pod Groups
+  can be complex.
+
+  - **Mitigation:** The initial version supports sequential processing of pods
+    within a PodGroup, avoiding complex backtracking or parallel processing
+    in the alpha release.
+
+## Design Details
+
+### Workload API Changes
+
+The Workload API (KEP-4671) will be extended to allow specifying group-level
+scheduling constraints. An optional `ScheduleConstraints` field is added to the
+`PodGroup` spec.
+
+```go
+// PodGroup (definition from KEP-4671, with additions)
+type PodGroup struct {
+    Name *string
+
+    // SchedulingConstraints defines group-level scheduling requirements,
+    // including topology and DRA colocation.
+    SchedulingConstraints *PodGroupSchedulingConstraints
+}
+
+// PodGroupSchedulingConstraints holds the scheduling constraints for the PodGroup.
+type PodGroupSchedulingConstraints struct {
+    // TopologyConstraints specifies desired topological placements for all pods
+    // within this PodGroup.
+    TopologyConstraints []TopologyConstraint
+
+    // DRAConstraints specifies constraints on how Dynamic Resources are allocated
+    // across the PodGroup.
+    DRAConstraints []DRAConstraint
+}
+
+// TopologyConstraint describes a desired topological colocation for all pods in the PodGroup.
+type TopologyConstraint struct {
+    // Level specifies the key of the node label representing the topology domain.
+    // All pods within the PodGroup must be colocated within the same domain instance.
+    // Examples: "topology.kubernetes.io/rack"
+    Level string
```

Comment:
Let's specify what happens if a PodGroup is replicated (there are multiple PodGroup instances/replicas).
I'm assuming that for each of those domain at that level is chosen, but there is not coordination between those (i.e. different pod group instances may be scheduled in different domains, but they can also share them).